tensorflow/core/common_runtime/zen_layout_pass.cc

```diff
@@ -0,0 +1,1292 @@
+  static bool CheckValidityForDTypeSupported(const Node *n) {
+    DataType data_type;
+    TF_CHECK_OK(GetNodeAttr(n->def(), "T", &data_type));
+    return (data_type == DT_FLOAT);
+  }
+
+  // Method to provide a 'valid' status for nodes that don't require any check.
+  // This method is used in ZenLayoutRewritePass() for creating the
+  // record/entry for rewriting native ops with Zen ops.
+  static bool RewriteValid(const Node *n) { return true; }
+
+  // Method to find whether the graph has inference ops only. It returns
+  // error status if the graph has training ops.
+  Status AreAllInferenceOps(std::unique_ptr<Graph> *g);
+
+  // Rewrites input node to a new node specified by its matching rewrite record.
+  //
+  // Input node may be deleted in case of rewrite. Attempt to use the node
+  // after the call can result in undefined behaviors.
+  //
+  // @input  g - input graph, n - Node to be rewritten,
+  //         ri - matching rewrite record,
+  //         reorder_flags - flags to populate reorder attributes of Zen op.
+  // @return OkStatus(), if the input node is rewritten;
+  //         Returns appropriate Status error code otherwise.
+  //         Graph is updated in case the input node is rewritten.
+  //         Otherwise, it is not updated.
+  Status ZenOpNodeRewrite(std::unique_ptr<Graph> *g, Node *n,
+                          const ZenOpRewriteRecord *ri,
+                          std::pair<bool, bool> reorder_flags);
+
+  // Functions specific to operators to copy attributes
+  // We need operator-specific function to copy attributes because the framework
+  // does not provide any generic function for it.
+  static void UpdateZenOpAttrs(const Node *orig_node, NodeBuilder *nb);
+
+  static void UpdateZenOpAttrsConv2D(const Node *orig_node, NodeBuilder *nb);
+
+  static void UpdateZenOpAttrsFusedConv2D(const Node *orig_node,
+                                          NodeBuilder *nb);
+
+  // Examines the input and output nodes of each node, for Zen nodes and
+  // determines reorder flags.
+  //
+  // @input   nodes - A list of nodes
+  // @return  An unordered map with nodes as key and
+  //          value as a pair of reorder flags
+  std::unordered_map<Node *, std::pair<bool, bool>> GetReorderFlags(
+      std::vector<Node *> &nodes);
+
+  // Update reorder information of all Zen nodes
+  //
+  // @input g - input graph
+  // @return true, if one or more updates are successful; false otherwise.
+  bool AddReorderAttrs(std::unique_ptr<Graph> *g);
+};
+
+// ZenLayoutRewritePass is executed in phase 0, to make sure it is executed
+// before MklLayoutRewritePass(Phase 1).
+REGISTER_OPTIMIZATION(OptimizationPassRegistry::POST_PARTITIONING, 0,
+                      ZenLayoutRewritePass);
+
+void DeleteNodeAndUpdateLinks(std::unique_ptr<Graph> *, Node *, Node *, int);
+
+const ZenLayoutRewritePass::ZenOpRewriteRecord *
+ZenLayoutRewritePass::CheckNodeForZenOpRewrite(const Node *n) const {
+  CHECK_NOTNULL(n);
```

Comment:
Fixed. Can you please let us know the reason behind this suggestion?

Reply:
Normally we don't allow new `CHECK`s in the code anymore. Our code checker won't let it pass unless we explicitly annotate it this way.

(We prefer `DCHECK` as an alternative so programs don't just crash. But `DCHECK` is also bad in a way that it gets optimized out for non-debug builds, and I don't think anyone builds TF in debug mode, so it would never catch anything either. So I'm fine with `CHECK`s in this PR, although I might come back to change them later.)

Reply:
Okay. Thank you for the clarification.